### PR TITLE
itest: Extend custom channel liquidity itest for RFQ HTLC tracking

### PR DESCRIPTION
**Pull request by GeorgeTsagk**
## Descriptions

Adds an extra edge case to the custom channels itest which verifieds that the rfq htlc tracking mechanism works as expected.

Related tapd PR: https://github.com/lightninglabs/taproot-assets/pull/1186

### CODE REVIEW

**Comment by ffranr** [comment]
@GeorgeTsagk do I only need to review the last three commits?

**Comment by GeorgeTsagk** [comment]
@ffranr this was just rebased, it's just one commit now (if you skip the temp one)

File: itest/litd_custom_channels_test.go
Comment (ffranr) [inline itest/litd_custom_channels_test.go:2125]: We have `BeforeTimeout` as a helper function. Would be good to have `AfterTimestamp` perhaps. But I think this sort of timing analysis strategy should be events driven instead.

File: itest/litd_custom_channels_test.go
Comment (ffranr) [inline itest/litd_custom_channels_test.go:2189]: needs to be removed?

File: itest/litd_custom_channels_test.go
Comment (ffranr) [inline itest/litd_custom_channels_test.go:2199]: I think you should use this patter here:
```
stream, err := erin.RouterClient.SendPaymentV2(ctxc, sendReq)
require.NoError(t.t, err)

_, _ = getPaymentResult(stream)
```

Need error check on `getPaymentResult` result?

File: itest/litd_custom_channels_test.go
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go:2082]: ah yes, will either update comment or value in the next push

i think i made it `/ 3` just to wait a bit less overall on the test

File: itest/litd_custom_channels_test.go
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go:2098]: This is a bit tricky. Under the hood we know that some HTLCs of some size are going to be forwarded and eventually cancelled. If we were to intercept (which adds some overhead in code complexity) we would also run into the issue of not knowing what to do with the HTLC we just intercepted. The reason the time based approach is nice here is because it's agnostic to the internal mechanisms of sender/router/receiver.

We don't know what's going on inside, but we want:
a) Sender to try to send payment and have their attempts failing
b) Slosh payment to complete ~halfway through the payment timeout
c) Sender to eventually complete payment

We could intercept HTLCs to guarantee that any attempt was made during a), but other than that could we use it any different way?

File: itest/litd_custom_channels_test.go
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go:2125]: thanks, will look into `BeforeTimeout` and see if it can save us some lines

File: itest/litd_custom_channels_test.go
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go:2189]: yeap leftover, ty

File: itest/litd_custom_channels_test.go
Comment (ffranr) [reply itest/litd_custom_channels_test.go:2098]: You call into `payInvoiceWithAssets` below this point. And internally that handles a stream:
https://github.com/lightninglabs/lightning-terminal/blob/e52a8c563321f4fcc71c0e5bb22da56f0531a6dc/itest/assets_test.go#L888

Could you perhaps call tapd `SendPayment` in your test and then monitor the stream?

I think that the stream should have a payment result payment status of in-flight before sloshing. And then after sloshing you can listen on the stream for a payment result payment status success.

That way you can be sure that the send attempt is in-flight and then is successful after sloshing. And without timing complexity or perhaps any extra goroutines?

Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go]: After latest foce-push I'm using a new approach, so PTAL

File: itest/litd_custom_channels_test.go
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go]: deprecated

File: itest/litd_custom_channels_test.go
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go]: not needed anymore with new approach

File: itest/litd_custom_channels_test.go
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go:2210]: latest approach uses existing helpers, closing

File: itest/litd_custom_channels_test.go
Comment (guggero) [inline itest/litd_custom_channels_test.go:1812]: Why do we need to force close here now?

File: itest/litd_custom_channels_test.go
Comment (guggero) [inline itest/litd_custom_channels_test.go:2098]: No need to manually sleep here I think. `assertNumHtlcs` will wait in a loop until the required number is reached.

File: itest/litd_custom_channels_test.go
Comment (guggero) [inline itest/litd_custom_channels_test.go:2135]: nit: formatting of multi-line function call.

File: itest/litd_custom_channels_test.go
Comment (guggero) [inline itest/litd_custom_channels_test.go:2203]: Shouldn't we instead set this to a way shorter timeout and then just expect the payment to fail?
Then we don't need the extra goroutine.

File: itest/litd_custom_channels_test.go
Comment (guggero) [inline itest/litd_custom_channels_test.go:2138]: We don't actually create a HODL invoice here. So we only test adding HTLCs to the counter in the order manager. But we don't test that if we cancel a HODL invoice that we then can make a normal payment again, as the liquidity isn't used up yet.

**Review by guggero — COMMENTED** [review]
Needs a couple of updates still, but definitely the better approach!

File: itest/litd_custom_channels_test.go
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go:1812]: The very last edge case leaves a payment hanging (i.e waiting for timeout) so the channel would not coop close as there were still in-flight HTLCs. I'll see if there's a way to bypass this and wait for the htlcs to get cancelled

File: itest/litd_custom_channels_test.go
Comment (GeorgeTsagk) [reply itest/litd_custom_channels_test.go:2138]: In the lines above we attempted to make a payment using multiple HTLCs (in the `assertNumHtlcs` calls above) using the same rfq scid

Since we asserted that HTLCs exist, then re-asserted that HTLCs do not exist anymore, that means the HTLCs got cancelled

am i missing smth?

Comment (guggero) [reply itest/litd_custom_channels_test.go:2144]: Hmm, I must've looked at an old version of the code or didn't see the `CancelInvoice` above. Looks good to me the way it is now.

**Review by guggero — APPROVED** [review]
LGTM pending dependent PRs.

File: itest/litd_custom_channels_test.go
Comment (ffranr) [inline itest/litd_custom_channels_test.go:2095]: ... providing the RFQ quote ID...

File: itest/litd_custom_channels_test.go
Comment (ffranr) [inline itest/litd_custom_channels_test.go:2138]: ... manually specified RFQ quote ID.

**Review by ffranr — APPROVED** [review]

File: itest/assets_test.go
Comment (guggero) [inline itest/assets_test.go:901]: Just use `i` here. That's the whole point of `WhenSome()`, that you don't have to use `UnsafeFrom()`.

**Comment by GeorgeTsagk** [comment]
Will wait for base PR to merge, then will update `go.mod` before merging this

File: itest/assets_test.go
Comment (guggero) [inline itest/assets_test.go:901]: `copy()` is tricky if the target slice is nil or empty, nothing happens. So you need to do `rfqBytes = make([]byte, len(i[:]))` first.

Comment (GeorgeTsagk) [reply itest/assets_test.go:901]: Nice catch! Interestingly enough the tests did not fail as the daemon would automatically acquire fresh quotes for the payment. Specifying an rfqID here was a effectively a noop.

The reason this wasn't caught by the next test case (which requires you use a specific rfq scid and none else) is because it is enforced by the receiver, while crafting the invoice.

**Pull request merged**
